File: src/platform/weather.ts
import { call } from './invoke';
import type {
  GeocodeResult,
  IpCityInfo,
  MsnOverview,
  WeatherCity,
  WeatherQueryOptions,
  WeatherResult,
} from './types';

export const weatherApi = {
  ipCity: (lang: string) => call<IpCityInfo>('weather_ip_city', { lang }),
  geocode: (city: string, lang: string) => call<GeocodeResult>('weather_geocode', { city, lang }),
  /** 服务端限速每秒一次 只在用户提交时调 */
  searchCity: (query: string, lang: string) => call<WeatherCity[]>('weather_search_city', { query, lang }),
  msnOverview: (lat: number, lon: number, locale: string) =>
    call<MsnOverview>('weather_msn_overview', { lat, lon, locale }),
  query: (lat: number, lon: number, opts?: WeatherQueryOptions) =>
    call<WeatherResult>('weather_query', {
      lat,
      lon,
      daily: opts?.daily ?? null,
      forecastDays: opts?.forecastDays ?? null,
    }),
};

File: src-tauri/src/ipc/weather.rs
use island_core::IpCityInfo;

use crate::error::AppResult;
use crate::services;

use super::off_thread;

#[tauri::command]
pub async fn weather_ip_city() -> AppResult<IpCityInfo> {
    off_thread(services::weather::ip_city).await
}

#[tauri::command]
pub async fn weather_geocode(city: String, lang: String) -> AppResult<serde_json::Value> {
    off_thread(move || services::weather::geocode(&city, &lang)).await
}

#[tauri::command]
pub async fn weather_query(
    lat: f64,
    lon: f64,
    daily: Option<String>,
    forecast_days: Option<u32>,
) -> AppResult<serde_json::Value> {
    off_thread(move || services::weather::query(lat, lon, daily.as_deref(), forecast_days)).await
}

#[tauri::command]
pub async fn weather_msn_overview(
    lat: f64,
    lon: f64,
    locale: String,
) -> AppResult<serde_json::Value> {
    off_thread(move || services::weather::msn_overview(lat, lon, &locale)).await
}
